frame/balances/src/lib.rs

```diff
@@ -944,6 +945,51 @@ impl<T: Config<I>, I: 'static> fungible::Unbalanced<T::AccountId> for Pallet<T,
 	}
 }
 
+impl<T: Config<I>, I: 'static> fungible::InspectReserve<T::AccountId> for Pallet<T, I> {
+	fn reserved_balance(who: &T::AccountId) -> T::Balance {
+		Self::account(who).reserved
+	}
+	fn total_balance(who: &T::AccountId) -> T::Balance {
+		Self::account(who).total()
+	}
+	fn can_reserve(who: &T::AccountId, amount: T::Balance) -> bool {
+		let a = Self::account(who);
+		let min_balance = T::ExistentialDeposit::get().max(a.frozen(Reasons::All));
+		if a.reserved.checked_add(&amount).is_none() { return false }
+		// We require it to be min_balance + amount to ensure that the full reserved funds may be
+		// slashed without compromising locked funds or destroying the account.
+		let required_free = match min_balance.checked_add(&amount) {
+			Some(x) => x,
+			None => return false,
+		};
+		a.free >= required_free
+	}
+}
+impl<T: Config<I>, I: 'static> fungible::MutateReserve<T::AccountId> for Pallet<T, I> {
+	fn reserve(who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		if amount.is_zero() { return Ok(()) }
+		ensure!(Self::can_reserve(who, amount), Error::<T, I>::InsufficientBalance);
```

Comment:
just a comment, that this error can also occur with overflow of `a.reserved.checked_add(&amount)`, so its not exactly always insufficient balance... but yeah, nit picks